numba/core/transforms.py

```diff
@@ -582,21 +702,104 @@ def within_known_range(start, end, known_ranges):
                 return True
         return False
 
-    setup_to_pop_block_map = {}
-    for s, e in sorted(find_ranges(blocks)):
-        # Look for setup_withs that have multiple pop_blocks each with a
-        # different target.
-        if s not in setup_to_pop_block_map:
-            setup_to_pop_block_map[s] = e
-        else:
-            if setup_to_pop_block_map[s] != e:
-                raise errors.CompilerError(
-                    "unsupported control flow: with-context contains branches "
-                    "(i.e. break/return/raise) that can leave the with-context. "
-                )
-        # Eliminate all withs contained within withs, we are only interested in
-        # the outermost with statements so we can lift them.
+    for s, e in sorted(with_ranges):
         if not within_known_range(s, e, known_ranges):
             known_ranges.append((s, e))
 
     return known_ranges
+
+def consolidate_multi_exit_withs(withs: dict, blocks, func_ir):
+    out = []
+    for k in withs:
+        vs : set = withs[k]
+        if len(vs) > 1:
+            func_ir, common = _fix_multi_exit_blocks(
+                func_ir, vs, split_condition=is_pop_block,
+            )
+            withs[k] = {common}
+    return func_ir
+
+
+def _fix_multi_exit_blocks(func_ir, exit_nodes, *, split_condition=None):
+    blocks = func_ir.blocks
+
+    any_blk = min(func_ir.blocks.values())
+    scope = any_blk.scope
+    max_key = max(func_ir.blocks) + 1
+
+    common_block = ir.Block(any_blk.scope, loc=any_blk.loc)
+    common_key = max_key
+    max_key += 1
+    blocks[common_key] = common_block
+
+    post_block = ir.Block(any_blk.scope, loc=any_blk.loc)
+    post_key = max_key
+    max_key += 1
+    blocks[post_key] = post_block
+
+    remainings = []
+    for i, k in enumerate(exit_nodes):
+        blk = blocks[k]
+
+        if split_condition is not None:
+            for pt, stmt in enumerate(blk.body):
+                if split_condition(stmt):
+                    break
+        else:
+            pt = -1
+
+        before = blk.body[:pt]
+        after = blk.body[pt:]
+        remainings.append(after)
+
+        blk.body = before
+        loc = blk.loc
+        blk.body.append(ir.Assign(value=ir.Const(i, loc=loc), target=scope.get_or_define("$cp", loc=loc), loc=loc))
```

Comment:
Suggest pulling out the name `"$cp"` as var and using that throughout, including things like e.g. `f"{the_cp_var}_rhs"` later. (aside: what's `cp` stand for here?)

Reply:
Also, long line!

Reply:
`cp` is control-point. as per https://dl.acm.org/doi/10.1145/2693261

Reply:
i wanted to have a prefix so we know where that variable is coming from for debugging.